### DISCUSSION TITLE
ProFTPD FTPS and SecureSocket

### DISCUSSION THREAD

**Discussion by pol2095** [airsdk/Adobe-Runtime-Support#3063] [Q&A] [answered]
Hello,

I try to connect to a ProFTPD server using [SecureSocket](https://help.adobe.com/fr_FR/FlashPlatform/reference/actionscript/3/flash/net/SecureSocket.html) 
```
var s:SecureSocket = new SecureSocket();
s.connect("mysite.com", 21);
```
but I have this message error :
`2031:Error #2031: Socket Error. URL: mysite.com serverCertificateStatus:invalid`

if I modify "tls.conf" in ProFTPD server adding `TLSOptions UseImplicitSSL NoSessionReuseRequired`, I can connect to ProFTPD server,
but I send the command :
```
s.writeUTFBytes("EPSV\r\n");
s.flush();
```
ProFTPD return an aleatory data port between 49152 and 65535 :
I try to connect to the port returned :
```
var r:SecureSocket = new SecureSocket();
r.connect("mysite.com", 51321);
```
but I have this messag error :
`2031:Error #2031: Socket Error. URL: mysite.com serverCertificateStatus:invalid`

I tried using Filezilla, explicit or implicit FTPS works correctly.

Using FTP (`new Socket();`), I can connect to the server using the data port returned by `EPSV` command but it isn't secure.

Thanks

**Comment by pol2095** [comment, marked as answer]
The data channel has a strange behaviour using TLS, it's neccesary to send the ftp command before to connect.

I have others questions :
I can't get explicit mode to work, explicit mode need to connect and send the ftp command `AUTH TLS` without TLS and after use TLS in the same socket session ?

**Comment by dyxribo** [comment]
does anyone know why this is happening? i think i have the same issue. 

i'm writing a helper class for general https connections, but i can't connect to anything due to the same error, not even google.

any help on what i could do or what is causing this would be greatly appreciated.

i'm using vscode on windows 11, with the newest stable SDK.

**Reply by ajwfrost** [reply]
Can you please provide a code snippet or similar? The below quick test worked fine for me:
```
			var secSocket : SecureSocket = new SecureSocket();
			secSocket.addEventListener(Event.CONNECT, function(e : Event) : void {
				trace("CONNECTED");
				secSocket.close();
			});
			trace("CONNECTING");
			secSocket.connect("www.google.com", 443);
```

**Reply by dyxribo** [reply]
no way... hours of messing with this and it turns out that i had to omit `https://` 😭

what i had before: 

```
_socket.connect("https://google.com", 443);
```

removed it and it worked instantly. i'm going to add a check to my helper class to remove it automatically in the future 😅

i'm a happy person again, thank you!

**Reply by ajwfrost** [reply]
Ah ... I feel that this is something we could flag up if a string argument to `Socket.connect` has a URL scheme (or we could just ignore that and pick up the host name...)!

**Reply by dyxribo** [reply]
absolutely, i just implemented the ignore in my helper class 😅 I'm actually surprised it wasn't doing that already. the docs don't mention it, but i guess it doesn't use it in the example either, so maybe that was an oversight on my part. 

not sure how many people are using SecureSocket and how often, but the issue is easy enough to work around, so i think it would be helpful at least to mention it in the docs. i could maybe submit a pull request to add a little note?
